### PR TITLE
Introduce `UberJarMergedResourceBuildItem` and `UberJarIgnoredResourceBuildItem`

### LINKED ISSUE #5677: shade-like transformers for uberjar creation
**Description**
When creating an uberjar with Quarkus, there is no way to deal with duplicated resources in the dependencies. The first resource found will be used and any other subsequent resource will be omitted. 

This is a typical problem when working with spring or many other frameworks that have duplicate entries under `/src/main/resources/META-INF` in the different .jar files that compose the framework.

**Implementation ideas**
The idea would be to
a) Try to integrate with maven's shade plugin. This plugin already has a built-in mechanism to deal with duplicated resources.
b) If the integration with shade is not possible, it would be nice to at least support a similar mechanism to declare `transformers` for files in the `quarkus-maven-plugin`.

### PR DESCRIPTION
**Pull request by gastaldi**
Fixes #5677

### CODE REVIEW

**Review by gsmet — DISMISSED** [review]
I haven't looked at it in detail but it would be better to name it `UberJarMergedResourceBuildItem`.

The way it is right now, it looks like an event that is triggered once resources are merged.

**Review by geoand — COMMENTED** [review]
It would be nice to have some kind of test for this

**Comment by gastaldi** [comment]
> It would be nice to have some kind of test for this

Definitely! Can you point me to a test that uses Build Items  for reference?

**Comment by gastaldi** [comment]
More specifically, this test doesn't work:

```java
package io.quarkus.deployment.pkg.builditem;

import java.io.IOException;
import java.net.URL;
import java.util.Collections;
import java.util.Enumeration;
import java.util.function.Consumer;

import io.quarkus.bootstrap.model.AppArtifact;
import io.quarkus.builder.BuildChainBuilder;
import io.quarkus.test.QuarkusUnitTest;
import org.jboss.shrinkwrap.api.ShrinkWrap;
import org.jboss.shrinkwrap.api.spec.JavaArchive;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.RegisterExtension;

import static org.assertj.core.api.Assertions.assertThat;

class UberJarMergedResourceBuildItemTest {

    @RegisterExtension
    static QuarkusUnitTest runner = new QuarkusUnitTest()
            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class))
            .withConfigurationResource("application-uber-jar.properties")
            .setForcedDependencies(
                    Collections.singletonList(
                            new AppArtifact("org.apache.cxf", "cxf-rt-transports-http", "3.4.3")
                    ))
            .addBuildChainCustomizer(buildCustomizer());

    @Test
    public void testResourceWasMerged() throws IOException {
        Enumeration<URL> resources = getClass().getClassLoader().getResources("META-INF/cxf/bus-extensions.txt");
        assertThat(resources.hasMoreElements()).isTrue();
        assertThat(resources.nextElement()).isNotNull();
        assertThat(resources.hasMoreElements()).isFalse();
    }

    static Consumer<BuildChainBuilder> buildCustomizer() {
        return (buildChainBuilder ->
                buildChainBuilder
                        .addBuildStep(context ->
                                              context.produce(new UberJarMergedResourceBuildItem("META-INF/cxf/bus-extensions.txt")))
                        .produces(UberJarMergedResourceBuildItem.class)
                        .build()
        );
    }
}
```

**Comment by geoand** [comment]
Ah darn, you are right.

We would likely need to add the same capability to the prod mode test.

**Comment by gastaldi** [comment]
Hm, `QuarkusProdModeTest` doesn't seem to work either. I'm adding my test to the `core/test-extension/deployment` because I can't add it to `core/deployment` (it would introduce a circular reference to quarkus-junit5-internal) but then it breaks other build steps 😛  
This is painful.

File: core/test-extension/deployment/src/test/java/io/quarkus/deployment/pkg/builditem/UberJarIgnoredResourceBuildItemTest.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        assertThat(runner.getExitCode()).isZero();
+    }
+
+    @QuarkusMain
+    public static class UberJarMain {
+
+        public static void main(String[] args) throws IOException {
+            List<URL> resources = Collections
+                    .list(UberJarMain.class.getClassLoader().getResources("META-INF/cxf/cxf.fixml"));
+            System.out.println("RESOURCES: " + resources.size());
```
Comment (geoand) [inline core/test-extension/deployment/src/test/java/io/quarkus/deployment/pkg/builditem/UberJarIgnoredResourceBuildItemTest.java:48]: Can we assert the content of the file?

**Review by gsmet — DISMISSED** [review]
The imports need a fix and the tests could be improved a bit. If this is done before tomorrow evening, I can backport it. Thanks!

**Review by geoand — APPROVED** [review]

**Pull request merged**